redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt

```diff
@@ -191,103 +200,162 @@ internal class UIViewBox :
+  get() {
+    var width = 0.0
+    var height = 0.0
+    var margin = Margin.Zero
+
+    forEachScoped { childModifier ->
+      when (childModifier) {
+        is Width -> with(Density.Default) {
+          width = maxOf(width, childModifier.width.toPx())
         }
 
-        Constraint.Wrap -> when (heightConstraint) {
-          Constraint.Fill -> { // Wrap Fill
-            for (subview in typedSubviews) {
-              subview.sizeThatFits(size).useContents {
-                maxItemWidth = maxOf(maxItemWidth, width)
-              }
-            }
-            maxItemHeight = size.useContents { height }
-          }
+        is Height -> with(Density.Default) {
+          height = maxOf(height, childModifier.height.toPx())
+        }
 
-          Constraint.Wrap -> { // Wrap Wrap
-            for (subview in typedSubviews) {
-              subview.sizeThatFits(size).useContents {
-                maxItemWidth = maxOf(maxItemWidth, width)
-                maxItemHeight = maxOf(maxItemHeight, height)
-              }
-            }
-          }
+        is Size -> with(Density.Default) {
+          width = maxOf(width, childModifier.width.toPx())
+          height = maxOf(height, childModifier.height.toPx())
         }
+
+        is ModifierMargin -> {
+          margin = maxEachSide(margin, childModifier.margin)
+        }
+      }
+    }
+
+    return CGSizeMake(width, height) + margin
+  }
+
+internal val Modifier.margin: Margin
+  get() {
+    var result = Margin.Zero
+
+    forEachScoped {
+      if (it is ModifierMargin) {
+        result = maxEachSide(result, it.margin)
       }
-      return CGSizeMake(
-        width = maxOf(maxRequestedWidth, maxItemWidth),
-        height = maxOf(maxRequestedHeight, maxItemHeight),
+    }
+
+    return result
+  }
+
+internal operator fun CValue<CGSize>.plus(margin: Margin): CValue<CGSize> {
+  return useContents {
+    with(Density.Default) {
+      CGSizeMake(
+        width = margin.start.toPx() + width + margin.end.toPx(),
+        height = margin.top.toPx() + height + margin.bottom.toPx(),
       )
     }
   }
 }
+
+internal val Margin.width: Double
+  get() = with(Density.Default) { start.toPx() + end.toPx() }
+
+internal val Margin.height: Double
+  get() = with(Density.Default) { top.toPx() + bottom.toPx() }
+
+/** Returns a size that takes the width from [widthSize] and the height from [heightSize]. */
+@Suppress("FunctionName")
+internal fun CGSizeMake(
+  widthSize: CValue<CGSize>,
+  heightSize: CValue<CGSize>,
+): CValue<CGSize> = CGSizeMake(
+  width = widthSize.useContents { width },
+  height = heightSize.useContents { height },
+)
+
+/** Returns the smallest size that wraps both [a] and [b]. */
+internal fun maxEachDimension(a: CValue<CGSize>, b: CValue<CGSize>): CValue<CGSize> {
+  return a@a.useContents {
+    b@b.useContents {
+      CGSizeMake(
+        maxOf(a@width, b@width),
+        maxOf(a@height, b@height),
+      )
+    }
+  }
+}
```

Comment:
This was the least worst of a few different ways of writing this function. Open to feedback!